x/staking/handler.go

```diff
@@ -229,7 +229,17 @@ func handleMsgDelegate(ctx sdk.Context, msg types.MsgDelegate, k keeper.Keeper)
 }
 
 func handleMsgUndelegate(ctx sdk.Context, msg types.MsgUndelegate, k keeper.Keeper) sdk.Result {
-	completionTime, err := k.Undelegate(ctx, msg.DelegatorAddress, msg.ValidatorAddress, msg.SharesAmount)
+	validator, found := k.GetValidator(ctx, msg.ValidatorAddress)
+	if !found {
+		return ErrNoValidatorFound(k.Codespace()).Result()
+	}
+
+	if validator.Tokens.IsZero() {
+		return ErrInsufficientShares(k.Codespace()).Result()
+	}
+
+	sharesAmt := validator.DelegatorShares.MulInt(msg.Amount.Amount).QuoInt(validator.Tokens).TruncateDec()
```

Comment:
Second point: this calculations needs to be consolidated with the other instance of this same calculation found here `https://github.com/cosmos/cosmos-sdk/blob/develop/x/staking/types/validator.go#L357` 

It should be added to a new function off of the validator:

The most appropriate place to add this calculation is right beside `ShareTokens` - this new function does the reverse operation, not sure on the best name though... maybe the new function should be called `SharesFromTokens`, and rename the old function to `TokensFromShares` (to reduce ambiguity) 

(p.s. this calculation should not be truncated call aka. `ShareTokensTruncated`)

Reply:
ACKed.